### PR TITLE
WIP: Decoding scripts using conformer LM.

### PR DESCRIPTION
**Pull request by csukuangfj**
Decoding code for https://github.com/k2-fsa/icefall/pull/54

Still working in progress. Hope to get some results this week.

### CODE REVIEW

**Comment by csukuangfj** [comment]
Here is the progress so far:

(1) Generate `masked_src`, `src`, `tgt`, `src_key_padding_mask`, and `weight` for the conformer lm model from
an nbest object

(2) Able to load the pre-trained conformer lm model into decode.py

Will add another option to decode.py to use conformer lm for rescoring.

----

Here is an example of generating the inputs for conformer lm from an nbest object.

Assume we have two utterances: utterance 0 has two paths and utterance 1 has 3 paths:
```python3
    # utterance 0
    path00 = k2.linear_fsa([1, 2, 0, 3, 0, 5])
    path01 = k2.linear_fsa([1, 0, 5, 0])

    # utterance 1
    path10 = k2.linear_fsa([9, 8, 0, 3, 0, 2])
    path11 = k2.linear_fsa([9, 8, 0, 0, 3, 2])
    path12 = k2.linear_fsa([9, 0, 8, 4, 0, 2, 3])

    shape = k2.RaggedShape("[[x x] [x x x]]")
```

The resulting inputs for conformer lm are:
(Assume bos_id=10, eos_id=20, blank_id=0, unmasked_weight=0.25.
For each pair of paths in an utterance, when one is used as a src, the other is used as a tgt.
)

### masked src
```
 tensor([[10,  1,  2,  3,  5, 20,  0],
        [10,  1,  0,  0,  5, 20,  0],
        [10,  1,  0,  0,  5, 20,  0],
        [10,  1,  5, 20,  0,  0,  0],
        [10,  9,  8,  3,  2, 20,  0],
        [10,  9,  8,  3,  2, 20,  0],
        [10,  9,  8,  0,  2,  0, 20],
        [10,  9,  8,  3,  2, 20,  0],
        [10,  9,  8,  3,  2, 20,  0],
        [10,  9,  8,  0,  2,  0, 20],
        [10,  9,  8,  0,  2,  0, 20],
        [10,  9,  8,  0,  2,  0, 20],
        [10,  9,  8,  4,  2,  3, 20]], dtype=torch.int32)
```

### src
```
 tensor([[10,  1,  2,  3,  5, 20,  0],
        [10,  1,  0,  0,  5, 20,  0],
        [10,  1,  2,  3,  5, 20,  0],
        [10,  1,  5, 20,  0,  0,  0],
        [10,  9,  8,  3,  2, 20,  0],
        [10,  9,  8,  3,  2, 20,  0],
        [10,  9,  8,  4,  2,  3, 20],
        [10,  9,  8,  3,  2, 20,  0],
        [10,  9,  8,  3,  2, 20,  0],
        [10,  9,  8,  4,  2,  3, 20],
        [10,  9,  8,  3,  2,  0, 20],
        [10,  9,  8,  3,  2,  0, 20],
        [10,  9,  8,  4,  2,  3, 20]], dtype=torch.int32)
```

### tgt
```
 tensor([[ 1,  2,  3,  5, 20,  0,  0],
        [ 1,  2,  3,  5, 20,  0,  0],
        [ 1,  0,  0,  5, 20,  0,  0],
        [ 1,  5, 20,  0,  0,  0,  0],
        [ 9,  8,  3,  2, 20,  0,  0],
        [ 9,  8,  3,  2, 20,  0,  0],
        [ 9,  8,  3,  2,  0, 20,  0],
        [ 9,  8,  3,  2, 20,  0,  0],
        [ 9,  8,  3,  2, 20,  0,  0],
        [ 9,  8,  3,  2,  0, 20,  0],
        [ 9,  8,  4,  2,  3, 20,  0],
        [ 9,  8,  4,  2,  3, 20,  0],
        [ 9,  8,  4,  2,  3, 20,  0]], dtype=torch.int32)
```

### src_key_padding_mask
```
 tensor([[False, False, False, False, False, False,  True],
        [False, False, False, False, False, False,  True],
        [False, False, False, False, False, False,  True],
        [False, False, False, False,  True,  True,  True],
        [False, False, False, False, False, False,  True],
        [False, False, False, False, False, False,  True],
        [False, False, False, False, False, False, False],
        [False, False, False, False, False, False,  True],
        [False, False, False, False, False, False,  True],
        [False, False, False, False, False, False, False],
        [False, False, False, False, False, False, False],
        [False, False, False, False, False, False, False],
        [False, False, False, False, False, False, False]])
```

### weight
```
 tensor([[0.2500, 0.2500, 0.2500, 0.2500, 0.2500, 0.0000, 0.0000],
        [0.2500, 1.0000, 1.0000, 0.2500, 0.2500, 0.0000, 0.0000],
        [0.2500, 1.0000, 1.0000, 0.2500, 0.2500, 0.0000, 0.0000],
        [0.2500, 0.2500, 0.2500, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.2500, 0.2500, 0.2500, 0.2500, 0.2500, 0.0000, 0.0000],
        [0.2500, 0.2500, 0.2500, 0.2500, 0.2500, 0.0000, 0.0000],
        [0.2500, 0.2500, 1.0000, 0.2500, 1.0000, 0.2500, 0.0000],
        [0.2500, 0.2500, 0.2500, 0.2500, 0.2500, 0.0000, 0.0000],
        [0.2500, 0.2500, 0.2500, 0.2500, 0.2500, 0.0000, 0.0000],
        [0.2500, 0.2500, 1.0000, 0.2500, 1.0000, 0.2500, 0.0000],
        [0.2500, 0.2500, 1.0000, 0.2500, 1.0000, 0.2500, 0.0000],
        [0.2500, 0.2500, 1.0000, 0.2500, 1.0000, 0.2500, 0.0000],
        [0.2500, 0.2500, 0.2500, 0.2500, 0.2500, 0.2500, 0.0000]])
```

**Comment by danpovey** [comment]
For the weights, here, we only actually care about the weights in the masked regions, so those 0.25's can be set to 0.  They were introduced for training purposes (the task is nontrivial only because we randomize the symbols in the non-masked regions, with some small probability).

**Comment by csukuangfj** [comment]
Thanks, will change it.

**Comment by csukuangfj** [comment]
I am able to use a masked conformer lm model to rescore a Nbest object and combine the resulting score with
attention score and acoustic score.

Will do some tests to see how it affects the WER.

---

By the way, the `conformer_ctc/decode.py` and `icefall/decode.py` become increasingly complicated and
more difficult to understand now.  I think we need another refactoring for the decoding scripts.

File: icefall/decode.py [hunk 1]
```diff
@@ -889,3 +895,198 @@ def rescore_with_attention_decoder(
...
+        )
+    else:
+        expanded_memory_key_padding_mask = None
+
+    # remove axis corresponding to states.
+    tokens_shape = nbest.fsa.arcs.shape().remove_axis(1)
+    tokens = k2.RaggedTensor(tokens_shape, nbest.fsa.tokens)
+    tokens = tokens.remove_values_leq(0)
+
+    alignment = compute_alignment(tokens, nbest.shape)
```
Comment (csukuangfj) [inline icefall/decode.py:990]: Code for conformer LM rescoring starts here.

File: icefall/lm/rescore.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+# limitations under the License.
+
+"""
+This file contains rescoring code for NN LMs, e.g., conformer LM.
+
+Support an utterance has 3 paths:
+    (a, b, c)
+and we want to use a masked conformer LM to assign a likelihood to each path.
+
+The following shows the steps:
```
Comment (csukuangfj) [inline icefall/lm/rescore.py:24]: @danpovey 
This shows how the masked LM model is used to assign a score to each path.

File: icefall/decode.py [hunk 3]
```diff
@@ -889,3 +897,225 @@ def rescore_with_attention_decoder(
...
+                src_key_padding_mask,
+            )
+
+            # nll means negative log-likelihood
+            # ll means log-likelihood
+            tgt_ll = -1 * (tgt_nll * tgt_weights).sum(dim=-1)
+
+            tgt_ll_list.append(tgt_ll)
+
+        #  tgt_ll = tgt_ll_list[1] - tgt_ll_list[0] # wer: 2.61
```
Comment (danpovey) [inline icefall/decode.py:1046]: This is a little suspicious to me; seems to me it should be hyp - ref = [1] - [0].  If the sign is wrong I think we should determine where.

Comment (csukuangfj) [inline icefall/decode.py:1046]: Suppose there is only one utterance and there are three paths. Let us call the three paths "a", "b", "c".

`tgt_ll_list[0]` contains
- "ab_self", use "a" as src and its shifted version as "tgt", the alignment is computed using the pair (a, b), where a is used as a reference and b as hypothesis
- "ac_self"
- "ba_self"
- "bc_self"
- "ca_self"
- "cb_self"

That is `[ab_self, ac_self, ba_self, bc_self, ca_self, cb_self]`



`tgt_ll_list[1]` contains
`[ab_other, ac_other, ba_other, bc_other, ca_other, cb_other]`

`tgt_ll = tgt_ll_list[0] - tgt_ll_list[1]`, which is
```
[ab_self - ab_other, ac_self - ac_other, ba_self - ba_other, bc_self - bc_other, ca_self - ca_other, cb_self - cb_other]
```

`ragged_tgt_ll` is
```
[ [ab_self - ab_other, ac_self - ac_other], 
  [ba_self - ba_other, bc_self - bc_other], 
  [ca_self - ca_other, cb_self - cb_other]
]
```

`ab_other - ab_self` can be interpreted as 
```
log p(b) - log p (a)
```
But I found that `ab_self - ab_other` has a lower WER. @danpovey

File: icefall/lm/rescore.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+    (i) Compute the alignment between "a" and "b"
+    (ii) Use the computed alignment as `masked_src,`
+    (iii) Use "a" as "src" and its shifted version as "tgt" (of course, we need
+          to add bos and eos) and we can get a log-likelihood value (after
+          negating the negative log-likelihood). Let us call this value as
+          "ab_self"
+    (iv)  Use "b" as "src" and its shifted version as "tgt".
+          We can get another likelihood value, denoted as "ab_other"
+
+So for the path pair (a, b), (a, c), we can get the following log-likelihood
```
Comment (csukuangfj) [inline icefall/lm/rescore.py:42]:
```suggestion
So for the path pair (a, b), (a, c), (b, a), (b, c), (c, a), and (c, b), we can get the following log-likelihood
```

File: icefall/decode.py [hunk 5]
```diff
@@ -889,3 +893,206 @@ def rescore_with_attention_decoder(
...
+    if attention_scale is None:
+        attention_scale_list = [0.01, 0.05, 0.08]
+        attention_scale_list += [0.1, 0.3, 0.5, 0.6, 0.7, 0.9, 1.0]
+        attention_scale_list += [1.1, 1.2, 1.3, 1.5, 1.7, 1.9, 2.0]
+        attention_scale_list += [2.1, 2.2, 2.3, 2.5, 3.0, 4.0, 5.0]
+    else:
+        attention_scale_list = [attention_scale]
+
+    if masked_lm_scale is None:
+        masked_lm_scale_list = [0.01, 0.05, 0.08]
```
Comment (danpovey) [inline icefall/decode.py:1075]: Perhaps, now that there are so many scales, we should start pruning the scales to the more frequently seen values.

File: icefall/lm/rescore.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        )
+        map_tensor_list.append(map_tensor)
+
+    return torch.cat(map_tensor_list)
+
+
+def compute_alignment(
+    tokens: k2.RaggedTensor,
+    shape: k2.RaggedShape,
+) -> k2.Fsa:
```
Comment (danpovey) [inline icefall/lm/rescore.py:225]: I think it would be especially helpful to have some documentation for this, esp. about the input and output formats.  Would be easier to follow what the function is doing.  Same with make_hyp_to_ref_map.

File: icefall/decode.py (hunk 5 above)
Comment (csukuangfj) [reply icefall/decode.py:1075]: Yes, I agree. It takes hours to compute the final WERs with so many scales.
